File: backend/timeline/services/calendar_chunks.py
"""Month-chunk windows for progressive Calendar loading.

Chunking is a delivery/render concern. The canonical forecast still covers the
full selected range; these windows only describe which days to return or mount.
"""
from __future__ import annotations

from calendar import monthrange
from datetime import date, timedelta

# Ranges at or under this many days are returned as a single chunk.
SHORT_RANGE_DAYS = 62
DEFAULT_MONTHS_PER_CHUNK = 2


def _month_end(d: date) -> date:
    return date(d.year, d.month, monthrange(d.year, d.month)[1])


def _add_months(d: date, months: int) -> date:
    month_index = d.month - 1 + months
    year = d.year + month_index // 12
    month = month_index % 12 + 1
    day = min(d.day, monthrange(year, month)[1])
    return date(year, month, day)


def calendar_chunk_windows(
    start: date,
    end: date,
    as_of: date,
    *,
    months_per_chunk: int = DEFAULT_MONTHS_PER_CHUNK,
) -> list[tuple[date, date]]:
    """
    Split [start, end] into month chunks.

    Short ranges (about 30–60 days) are a single window.
    Longer ranges start with the as-of month plus the next month (clipped to
    the selected range), then continue in ``months_per_chunk`` steps.
    """
    if start > end:
        return []
    span_days = (end - start).days + 1
    if span_days <= SHORT_RANGE_DAYS:
        return [(start, end)]

    months_per_chunk = max(1, months_per_chunk)
    first_month = date(as_of.year, as_of.month, 1)
    first_end_month = _add_months(first_month, months_per_chunk - 1)
    first_end = min(end, _month_end(first_end_month))
    if first_end < start:
        first_end = min(end, _month_end(start))

    windows = [(start, first_end)]
    cursor = first_end + timedelta(days=1)
    while cursor <= end:
        last_month = _add_months(date(cursor.year, cursor.month, 1), months_per_chunk - 1)
        chunk_end = min(end, _month_end(last_month))
        windows.append((cursor, chunk_end))
        cursor = chunk_end + timedelta(days=1)
    return windows

File: apps/mobile/features/calendar/calendarChunks.ts
/** Progressive calendar chunk windows — aligned with backend ``calendar_chunks.py`` and web. */

export const SHORT_RANGE_DAYS = 62;
export const DEFAULT_MONTHS_PER_CHUNK = 2;

export type CalendarChunkWindow = {
  start: string;
  end: string;
};

function parseIso(iso: string): Date {
  const [y, m, d] = iso.split("-").map(Number);
  return new Date(y, m - 1, d);
}

function toIsoDate(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

function monthEnd(d: Date): Date {
  return new Date(d.getFullYear(), d.getMonth() + 1, 0);
}

function addCalendarMonths(d: Date, months: number): Date {
  const year = d.getFullYear();
  const month = d.getMonth() + months;
  const last = new Date(year, month + 1, 0).getDate();
  return new Date(year, month, Math.min(d.getDate(), last));
}

function addDays(d: Date, days: number): Date {
  const next = new Date(d.getTime());
  next.setDate(next.getDate() + days);
  return next;
}

export function calendarChunkWindows(
  startIso: string,
  endIso: string,
  asOfIso: string,
  monthsPerChunk: number = DEFAULT_MONTHS_PER_CHUNK
): CalendarChunkWindow[] {
  const start = parseIso(startIso);
  const end = parseIso(endIso);
  const asOf = parseIso(asOfIso);
  if (start > end) return [];
  const spanDays = Math.round((end.getTime() - start.getTime()) / 86400000) + 1;
  if (spanDays <= SHORT_RANGE_DAYS) {
    return [{ start: startIso, end: endIso }];
  }
  const step = Math.max(1, monthsPerChunk);
  const firstMonth = new Date(asOf.getFullYear(), asOf.getMonth(), 1);
  const firstEndMonth = addCalendarMonths(firstMonth, step - 1);
  let firstEnd = monthEnd(firstEndMonth);
  if (firstEnd > end) firstEnd = end;
  if (firstEnd < start) {
    const startMonthEnd = monthEnd(start);
    firstEnd = startMonthEnd > end ? end : startMonthEnd;
  }

  const windows: CalendarChunkWindow[] = [{ start: toIsoDate(start), end: toIsoDate(firstEnd) }];
  let cursor = addDays(firstEnd, 1);
  while (cursor <= end) {
    const lastMonth = addCalendarMonths(new Date(cursor.getFullYear(), cursor.getMonth(), 1), step - 1);
    let chunkEnd = monthEnd(lastMonth);
    if (chunkEnd > end) chunkEnd = end;
    windows.push({ start: toIsoDate(cursor), end: toIsoDate(chunkEnd) });
    cursor = addDays(chunkEnd, 1);
  }
  return windows;
}

export function chunkCoversMonth(
  chunk: CalendarChunkWindow,
  year: number,
  month: number
): boolean {
  const monthStart = `${year}-${String(month + 1).padStart(2, "0")}-01`;
  const last = toIsoDate(monthEnd(new Date(year, month, 1)));
  return chunk.start <= last && chunk.end >= monthStart;
}

export function chunkWindowForMonth(
  windows: CalendarChunkWindow[],
  year: number,
  month: number
): CalendarChunkWindow | undefined {
  return windows.find((window) => chunkCoversMonth(window, year, month));
}
